cuqi/experimental/mcmc/_gibbs.py

```diff
@@ -0,0 +1,217 @@
+        self._set_targets()
+
+        # Run over pre-sample methods for samplers that have it
+        # TODO. Some samplers (NUTS) seem to require to run _pre_warmup before _pre_sample
+        # This is not ideal and should be fixed in the future
+        for sampler in self.samplers.values():
+            self._pre_warmup_and_pre_sample_sampler(sampler)
+
+        # Validate all targets for samplers.
+        self.validate_targets()
+
+    # ------------ Public methods ------------
+    def validate_targets(self):
+        """ Validate each of the conditional targets used in the Gibbs steps """
+        if not isinstance(self.target, JointDistribution):
+            raise ValueError('Target distribution must be a JointDistribution.')
+        for sampler in self.samplers.values():
+            sampler.validate_target()
+
+    def sample(self, Ns) -> 'GibbsNew':
+        """ Sample from the joint distribution using Gibbs sampling """
+        for _ in progressbar(range(Ns)):
+            self.step()
+            self._store_samples()
+
+    def warmup(self, Nb) -> 'GibbsNew':
+        """ Warmup (tune) the Gibbs sampler """
+        for idx in progressbar(range(Nb)):
+            self.step_tune(idx)
+            self._store_samples()
+
+    def get_samples(self) -> Dict[str, Samples]:
+        samples_object = {}
+        for par_name in self.par_names:
+            samples_array = np.array(self.samples[par_name]).T
+            samples_object[par_name] = Samples(samples_array, self.target.get_density(par_name).geometry)
+        return samples_object
+    
+    def step(self):
+        """ Sequentially go through all parameters and sample them conditionally on each other """
+
+        # Sample from each conditional distribution
+        for par_name in self.par_names:
+
+            # Set targets (TODO: This is inefficient. Instead we should only update the target for the current parameter)
+            self._set_targets()
+
+            # Get sampler
+            sampler = self.samplers[par_name]
+
+            # Set initial point using current samples and reinitalize sampler
+            # This makes the sampler loose all of its state
+            # We need to design tests that allow samplers to change target
+            # and not require reinitialization. This is needed to keep properties
+            # like the internal state of NUTS for the next Gibbs step.
+            sampler.initial_point = self.current_samples[par_name]
+            sampler.reinitialize()
+
+            # Run pre_warmup and pre_sample methods for sampler
+            # TODO. Some samplers (NUTS) seem to require to run _pre_warmup before _pre_sample
+            self._pre_warmup_and_pre_sample_sampler(sampler)
+
+            # Take a MCMC step
+            sampler.step()
+
+            # Extract samples (Ensure even 1-dimensional samples are 1D arrays)
+            self.current_samples[par_name] = sampler.current_point.reshape(-1)
+
+    def step_tune(self, idx):
```

Comment:
It seems "regular" samplers have do not have a `step_tune` methods but only separate `step` and `tune` methods.  Can the Gibbs sampler also just have `step` and `tune` and no `step_tune`?

Reply:
Agreed. Fixed.